### PR TITLE
Improve TeamColorMode

### PR DESCRIPTION
**Pull request by Strogoo**
This requires an engine patch: https://github.com/FAForever/FA-Binary-Patches/pull/5
To test this pr you can compile exe from linked PR using FaP.exe

So what it does is allowing player to customize colors for team color mode. Originally this feature is completely useless as it uses one color for all allies and one for all enemies.

изображение

I'm planning to add an option "Change colors in scoreboard" later, so when you turn on team colors it will also change icon's colors in scoreboard (will be less confusing I guess, anyway it will be optional).

### CODE REVIEW

**Comment by Garanas** [comment]
I have not had the time to look / review this PR but I just wanted to say that this is great work - I'll look at it on Sunday. How did executable releases happen in the past? I know that KionX has a few changes ready to go too that have not gone live yet.

**Comment by Strogoo** [comment]
As far as I remember, someone who is responsible for an exe patches (first it was Kyryll, now it's Kionix, well basically the one who does most of asm job :) gives a patched exe to a fa repo admin (or mb to server guys directly? As we don't have an exe file in fa repo) so they can upload it manually. But I can be wrong

Also now we have an binary-patches repo and a patcher repo with released patcher in it, so anyone can make a patched exe in a few clicks and without any special knowledge

**Comment by Strogoo** [comment]
ok, now it changes scoreboard colors too (optional). Supports both the original scoreboard and scoreboard from SSB mod.

**Comment by Sheikah45** [comment]
As far as the binary patches and releases go we just need to create an endpoint in the API and then it will be a part of the binary_patches CI to update it, just have to work out the versioning and things like that.

**Comment by Garanas** [comment]
Just tried to give it a go:

```yaml
WARNING: Unable to find file /mods/supremescoreboard/modules/score_board.lua
WARNING: stack traceback:
WARNING:         [C]: in function `doscript'
WARNING:         [C]: ?
WARNING:         ...documents\supreme commander\fa\lua\system\import.lua(87): in function `import'
WARNING:         ...documents\supreme commander\fa\lua\system\import.lua(72): in function `import'
WARNING:         ...s\supreme commander\fa\lua\ui\game\multifunction.lua(1138): in function <...s\supreme commander\fa\lua\ui\game\multifunction.lua:1137>
WARNING:         ...s\supreme commander\fa\lua\ui\game\multifunction.lua(1180): in function <...s\supreme commander\fa\lua\ui\game\multifunction.lua:1155>
WARNING:         ...s\supreme commander\fa\lua\ui\game\multifunction.lua(1370): in function `TeamColorHandler'
WARNING:         ...s\supreme commander\fa\lua\ui\game\multifunction.lua(86): in function `OnClick'
WARNING:         ...alliance\gamedata\mohodata.scd\lua\maui\checkbox.lua(96): in function <...alliance\gamedata\mohodata.scd\lua\maui\checkbox.lua:75>
WARNING: Error running HandleEvent script in CScriptObject at 13f13c40: ...documents\supreme commander\fa\lua\system\import.lua(72): Error importing '/mods/supremescoreboard/modules/score_board.lua'
         stack traceback:
         	[C]: in function `error'
         	...documents\supreme commander\fa\lua\system\import.lua(92): in function `import'
         	...documents\supreme commander\fa\lua\system\import.lua(72): in function `import'
         	...s\supreme commander\fa\lua\ui\game\multifunction.lua(1138): in function <...s\supreme commander\fa\lua\ui\game\multifunction.lua:1137>
         	...s\supreme commander\fa\lua\ui\game\multifunction.lua(1180): in function <...s\supreme commander\fa\lua\ui\game\multifunction.lua:1155>
         	...s\supreme commander\fa\lua\ui\game\multifunction.lua(1370): in function `TeamColorHandler'
         	...s\supreme commander\fa\lua\ui\game\multifunction.lua(86): in function `OnClick'
         	...alliance\gamedata\mohodata.scd\lua\maui\checkbox.lua(96): in function <...alliance\gamedata\mohodata.scd\lua\maui\checkbox.lua:75>
```

It is now dependent on Supreme Score board 😄

**Comment by Strogoo** [comment]
Very interesting :) I tried it with SSB turned off, but the mod folder was still there, that's why it didn't crash for me. Ehh, I'll try to find another way.

I wish we have something similar to `try` in Python, when you just put it before some garbage trash non-tested code, and even if it will crash in some circumstances, everything goes smooth and code execution won't stop :D

**Comment by Garanas** [comment]
Try pcall: https://www.lua.org/pil/8.4.html

```
local status, err = pcall(function () error({code=121}) end)
print(err.code)  -->  121
```

**Comment by Strogoo** [comment]
Should be ok now. Tested in 3 conditions: 1. No SSB at all 2. SSB folder exists but mod is turned off 3. SSB on

File: lua/ui/game/multifunction.lua [hunk 1]
```diff
@@ -1135,7 +1135,12 @@ local function saveTeamColorsToPrefs()
 end
 
 local function changeScoreboardColors(colorsTbl)
-    local armyLines = import('/mods/SupremeScoreBoard/modules/score_board.lua').controls.armyLines or import('/lua/ui/game/score.lua').controls.armyLines
+    local armyLines
+    if exists('/mods/SupremeScoreBoard/modules/score_board.lua') then
```
Comment (speed2CZ) [inline lua/ui/game/multifunction.lua:1139]: This is wrong. The mod should be updated instead

Comment (Strogoo) [reply lua/ui/game/multifunction.lua:1139]: Why? I used an old version of SSB for my previous tests (downloaded latest version just today) and all worked fine. Ofc we can add a version controller here, but honeslty, I have zero motivation to do that :)

Also taking in account that SSB update rate is like 1 patch per 2 years or smth? And I doubt core code will be changed some day, so this `controls.armyLines` will be there until the end of days. So pls let it be like that, tyvm <3

Comment (Garanas) [reply lua/ui/game/multifunction.lua:1139]: I agree with Speed2 - the mod should be updated to support this new functionality over trying to 'include' the mod via the repository by having a check like this. Is @HUSSAR-mtrela around to do this?

Comment (Strogoo) [reply lua/ui/game/multifunction.lua:1139]: Well, If Hussar isn't afk then sure :) Anyway, for SSB we can do an exception and use this "dirty" (just a little bit) solution. This way it works with any version of mod and won't break anything

**Comment by Garanas** [comment]
I'll review this again tomorrow.

**Comment by KionX** [comment]
It looks broken.
There is no backward compatibility.
There is no reset option to FA settings.
The window closing button is poorly visible.

**Comment by Garanas** [comment]
I may redo the UI of this pull request, would that be alright with you?

**Comment by Strogoo** [comment]
@Garanas Yep, feel free to change anything.

**Comment by Garanas** [comment]
Closing this pull request as the changes to the engine required for it has been closed too.

**Comment by Strogoo** [comment]
New engine PR https://github.com/FAForever/FA-Binary-Patches/pull/8 (approved by KioniX in pm :)

**Comment by Garanas** [comment]
Closing until the engine patch has been merged.

**Comment by KionX** [comment]
Patch has been merged.

**Comment by KionX** [comment]
@Garanas, @Strogoo
Is anyone going to finish it?

**Comment by Garanas** [comment]
We'll have to before we do a release.

**Comment by Strogoo** [comment]
Ok, so KionX changed string parsing a bit:
```
TeamColorMode("ffffffff,ffff32ff,ffb76518,ffa79602")
```

so no more additional `(` `)` symbols, also it supports uppercase now. Just tested with new exe - all works fine.

The only problem is ugly UI :) I leave that for Jip (or you can just merge it as is, because this settings window won't be used that often anyway)

**Comment by Strogoo** [comment]
I remade UI so now it uses same bg as minimap settings window:
изображение

Further improvements (if needed) is on you :)

**Comment by Garanas** [comment]
As good as we'll have it for now - I'll make an issue that describes how to tackle this properly. Note to https://github.com/FAForever/fa/pull/3939, once complete: remove the hack for SSB.

**Pull request merged**